Fixed #36526 -- Batching bulk_update to reduce memory usage.

Jkhall81 opened the pull request:
#### Trac ticket number
https://code.djangoproject.com/ticket/36526

ticket-36526

#### Branch description
This PR address ticket #36526 by modifying `QuerySet.bulk_update()` to execute each batch immediately, rather than accumulating all update queries in memory and executing them at the end.

This change significantly reduces memory usage and improves efficiency for large updates.  In a benchmark test updating 100 million objects:
- Before patch: peak memory usage was ~1065 MB and test runtime was ~507 seconds.
- After patch: peak memory useage dropped to 8 MB and runtime was ~477 seconds.

The memory profile is now flat and stable even with massive updates, avoiding the previous linear growth.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

Jkhall81 commented:
CI shows failures in unrelated test modules (502 in total) such as:
- "absolute_url_overrides.tests.AbsoluteUrlOverrideTests
- "invalid_models_tests.test_models"

I just looked at these two.  These failures involve dynamically created models that lack an explicit app_label and are not listed in INSTALLED_APPS.

This patch only modifies QuerySet.bulk_update() logic and adds isolated tests in a dedicated tests/bulk_update/ folder.  It does not alter global model behavior, test settings, or installed apps.

The failures are isolated to specific jobs (e.g. MySQL and SpatialLite on Python 3.12/3.13).  I am not sure why they are showing up.

anze3db commented on tests/bulk_update/test_bulk_update_batching.py:
```diff
@@ -0,0 +1,35 @@
+    @classmethod
+    def setUpTestData(cls):
+        Thing.objects.bulk_create(
+            [Thing(name=f"Thing {i}", description="Old") for i in range(1_000_000)],
```
Creating a million objects is probably too slow to run in a test suite or was this just for debugging?

anze3db commented on .gitignore:
```diff
@@ -17,3 +17,4 @@ tests/.coverage*
 build/
 tests/report/
 tests/screenshots/
+tests/test_postgres.py
```
This doesn't feel related to the ticket, was it intentional?

Jkhall81 replied:
Yes.  I put that there so that file wouldn't get included in the PR.

Jkhall81 commented:
Removed inline comments from query.py, reduced the number of objects in the test, replaced the print() statement with an assert, and restored .gitignore to its original state.

charettes commented on django/db/models/query.py:
```diff
@@ -933,26 +933,27 @@ def bulk_update(self, objs, fields, batch_size=None):
         requires_casting = connection.features.requires_casted_case_in_updates
         batches = (objs[i : i + batch_size] for i in range(0, len(objs), batch_size))
-        updates = []
-        for batch_objs in batches:
-            update_kwargs = {}
-            for field in fields:
-                when_statements = []
-                for obj in batch_objs:
-                    attr = getattr(obj, field.attname)
-                    if not hasattr(attr, "resolve_expression"):
-                        attr = Value(attr, output_field=field)
-                    when_statements.append(When(pk=obj.pk, then=attr))
-                case_statement = Case(*when_statements, output_field=field)
-                if requires_casting:
-                    case_statement = Cast(case_statement, output_field=field)
-                update_kwargs[field.attname] = case_statement
-            updates.append(([obj.pk for obj in batch_objs], update_kwargs))
```
Unfortunately performing the resolving of the expression (which is known to take a significant amount of time see ticket-31202) while the transaction is opened is likely to be more damageable than the memory tax it incurs.
